### PR TITLE
Stale config cleanup commands

### PR DESCRIPTION
**Pull request by kenny-statsig**
Simple quick command:

https://user-images.githubusercontent.com/111380336/200081104-89acfa3b-2d13-49e7-91da-28476d517fd9.mov

Callable bulk cleanup:

https://user-images.githubusercontent.com/111380336/200438625-53652a55-fdb8-48c7-87c9-5e2ff9f7af5b.mov

### CODE REVIEW

File: src/providers/ConfigCodeActionProvider.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  title: string;
+  kind: vsc.CodeActionKind;
+};
+
+export const CODE_ACTIONS: Record<string, CodeActionType> = {
+  removeStaleConfig: {
+    command: 'remove-stale-config',
+    title: 'Cleanup',
+    kind: vsc.CodeActionKind.QuickFix,
+  },
```
Comment (tore-statsig) [inline src/providers/ConfigCodeActionProvider.ts:29]: I would make two actions for gates: replace with `true` or replace with `false`

File: src/providers/ConfigCodeActionProvider.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    if (getConfigMatch) {
+      getConfigMatch.forEach((match) => {
+        const matchIndex = line.text.indexOf(match);
+        const matchRange = new vsc.Range(
+          line.lineNumber,
+          matchIndex,
+          line.lineNumber,
+          matchIndex + match.length,
+        );
+        edit.delete(doc.uri, matchRange);
```
Comment (tore-statsig) [inline src/providers/ConfigCodeActionProvider.ts]: replace config checks with an empty object `{}`

**Review by tore-statsig — CHANGES_REQUESTED** [review]
Woah this is neat!

Instead of just removing it, for gates, lets replace with true/false

for configs/experiments, replace with an empty object.  People can clean up from there

File: src/lib/languageUtils.ts [hunk 3]
```diff
@@ -1,2 +1,29 @@
...
+  }
+}
+
+export function getStatsigAPIGetConfigRegex(
+  language: SupportedLanguageType,
+): RegExp {
+  switch (language) {
+    case 'javascript':
+    case 'typescript':
+      return /[a-zA-Z0-9_]+[.]+getConfig+\(+[^)]+\)/;
```
Comment (tore-statsig) [inline src/lib/languageUtils.ts]: `getConfig` or `getExperiment`, both return a `DynamicConfig`

Comment (kenny-statsig) [reply src/lib/languageUtils.ts]: Oh interesting, so the "dynamic_configs" from DCS can be referenced by both `getConfig` and `getExperiment`? Or do you mean they share the same base `Ent`

File: src/providers/ConfigCodeActionProvider.ts (hunk 1 above)
Comment (kenny-statsig) [reply src/providers/ConfigCodeActionProvider.ts:29]: Considering these are for gates that they probably forgot about, I feel like it might be more user-friendly to just not give them the option. What if we just replace with `false`?

File: src/providers/diagnostics.ts [hunk 4]
```diff
@@ -17,57 +17,63 @@ export function refreshDiagnostics(
...
   for (let lineIndex = 0; lineIndex < doc.lineCount; lineIndex++) {
     const lineOfText = doc.lineAt(lineIndex);
-    const configSearch = CONFIG_NAME_WITH_QUOTES_REGEX.exec(lineOfText.text);
+    const configSearch = lineOfText.text.match(CONFIG_NAME_WITH_QUOTES_REGEX);
     if (configSearch !== null) {
       configSearch.forEach((match) => {
         const configName = match.substring(1, match.length - 1);
-        getStaleConfig(configName).forEach((config) => {
-          diagnostics.push(
+        if (getStaleConfig(configName).length > 0) {
```
Comment (kenny-statsig) [inline src/providers/diagnostics.ts:35]: These changes are mostly just fixing an edge case where the variable name contains the actual config name as a substring (e.g. `const testGate = 'test'`)

File: src/providers/ConfigCodeActionProvider.ts [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    if (ConfigCodeActionProvider.handleEdit(doc, edit, staleConfigs)) {
+      staleConfigs.forEach((config) => {
+        configCounts[config]
+          ? ++configCounts[config]
+          : (configCounts[config] = 1);
+      });
+      cleanedFiles.push(file.path);
+      output.appendLine(`(success) ${file.path}`);
+    } else {
+      output.appendLine(`(failed) ${file.path}`);
```
Comment (kenny-statsig) [inline src/providers/ConfigCodeActionProvider.ts:192]: This case is specifically if we detect usage of a stale gate but cannot actually remove it (demonstrated in test plan). This happens when we don't have coverage of a certain syntax where the config can be used.

File: src/providers/ConfigCodeActionProvider.ts (hunk 1 above)
Comment (jacob-statsig) [reply src/providers/ConfigCodeActionProvider.ts:29]: Is it possible a user could forget about a gate that always returns true, such as, after a rollout of a new feature has completed? In which case `false` isn't ideal.

Comment (tore-statsig) [reply src/providers/ConfigCodeActionProvider.ts:29]: I think we will want both (you can disable a gate and set true or false as the default value), but if you want to just start with false thats fine.

I suppose it kind of depends on the context - if the gate hasnt been checked in a while, false makes the most sense.  If the gate has been at 100% for a while, true makes the most sense.

Right now I think your staleconfigs are only based on checks in last 30 days, so false is best

File: src/lib/languageUtils.ts [hunk 6]
```diff
@@ -1,2 +1,77 @@
...
+
+export function getStatsigAPIGetConfigRegex(
+  language: SupportedLanguageType,
+  configName: string,
+): RegExp {
+  switch (language) {
+    case 'javascript':
+    case 'typescript':
+      return new RegExp(
+        `[a-zA-Z0-9_]+[.]getConfig\\([^)]*["'\`]${configName}["'\`][^)]*\\)`,
```
Comment (tore-statsig) [inline src/lib/languageUtils.ts:34]: can you add getExperiment too

File: src/lib/languageUtils.ts [hunk 7]
```diff
@@ -1,2 +1,77 @@
...
+  return (SUPPORTED_LANGUAGES as unknown as string[]).includes(language);
+}
+
+export function getStatsigAPICheckGateRegex(
+  language: SupportedLanguageType,
+  configName: string,
+): RegExp {
+  switch (language) {
+    case 'javascript':
+    case 'typescript':
```
Comment (jacob-statsig) [inline src/lib/languageUtils.ts:18]: Would it be easier to have a `supportedLanguage` interface with all the functions in this file? And then all your new languages could just inherit

File: src/lib/languageUtils.ts (hunk 3 above)
Comment (tore-statsig) [reply src/lib/languageUtils.ts]: Well, both.  getExperiment and getConfig do mostly the same things (both return a DynamicConfig and they are mostly interchangeable, there are some special parameters for getExperiment in client SDKs (sticky experiments).

Can you add getExperiment? Still didnt see that

**Review by tore-statsig — APPROVED** [review]
awesome.  Add getExperiment to the regex for sure

File: src/lib/languageUtils.ts (hunk 7 above)
Comment (kenny-statsig) [reply src/lib/languageUtils.ts:18]: Yep definitely. Will leave this restructuring to when I get around to supporting multiple languages.
